crates/extensions/tedge_config_manager/src/actor.rs

```diff
@@ -516,6 +548,142 @@ impl ConfigManagerActor {
+        .with_context(|| format!("failed to open temporary source file '{src}'"))?;
+
+    // TODO: create tests to ensure writes we expect are atomic
+    let mut tempfile = tempfile::Builder::new()
+        .permissions(std::fs::Permissions::from_mode(0o600))
+        .tempfile_in(dest.parent().context("invalid path")?)
+        .with_context(|| format!("could not create temporary file at '{dest}'"))?;
+
+    std::io::copy(&mut src_file, &mut tempfile).context("failed to copy")?;
+
+    tempfile
+        .as_file()
+        .set_permissions(std::fs::Permissions::from_mode(target_permissions.mode))
+        .context("failed to set mode on the destination file")?;
+
+    fchown(
+        tempfile.as_file(),
+        Some(target_permissions.uid),
+        Some(target_permissions.gid),
+    )
+    .context("failed to change ownership of the destination file")?;
+
+    tempfile.as_file().sync_all()?;
+
+    tempfile
+        .persist(dest)
+        .context("failed to persist temporary file at destination")?;
+
+    Ok(())
+}
+
+/// Computes target permissions for deployment of the config file.
+///
+/// - if file exists preserve current permissions
+/// - if it doesn't exist apply permissions from `permissions` if they are defined
+/// - set to root:root with default umask otherwise
+///
+/// # Errors
+/// - if desired user/group doesn't exist on the system
+/// - no permission to read destination file
+fn config_file_target_permissions(
+    file_entry: &FileEntry,
+    dest: &Utf8Path,
+) -> anyhow::Result<Permissions> {
+    let current_file_permissions = match std::fs::metadata(dest) {
+        Err(err) => match err.kind() {
+            ErrorKind::PermissionDenied => return Err(err).context("no permissions"),
+            ErrorKind::NotFound => None,
+            _ => return Err(err).context("unexpected IO error"),
+        },
+        Ok(p) => Some(p),
+    };
+
+    let entry_uid = if let Some(ref u) = file_entry.file_permissions.user {
+        let uid = uzers::get_user_by_name(u)
+            .with_context(|| format!("no such user: '{u}'"))?
+            .uid();
+        Some(uid)
+    } else {
+        None
+    };
+
+    let entry_gid = if let Some(ref g) = file_entry.file_permissions.group {
+        let gid = uzers::get_group_by_name(g)
+            .with_context(|| format!("no such group: '{g}'"))?
+            .gid();
+        Some(gid)
+    } else {
+        None
+    };
+    let entry_mode = file_entry.file_permissions.mode;
+
+    let uid = current_file_permissions
+        .as_ref()
+        .map(|p| p.uid())
+        .or(entry_uid)
+        .unwrap_or(0);
+
+    let gid = current_file_permissions
+        .as_ref()
+        .map(|p| p.gid())
+        .or(entry_gid)
+        .unwrap_or(0);
+
+    let mode = current_file_permissions
+        .as_ref()
+        .map(|p| p.mode())
+        .or(entry_mode)
+        .unwrap_or(0o644);
+
+    Ok(Permissions { uid, gid, mode })
+}
+
+struct Permissions {
```

Comment:
Why not reuse `tedge_utils::file::PermissionEntry`?

Reply:
https://github.com/thin-edge/thin-edge.io/pull/3069#discussion_r1736017001

We want to use uid/gid directly instead of names, since doing so avoids an extra failure mode.